### PR TITLE
Lots of good things

### PR DESCRIPTION
**Pull request by chrilves**
Precepte is now

- Stack safe
- A proper monad state
- A monad error
- able to use Algebraic Effects

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -54,7 +70,7 @@ lazy val commonSettings =  Seq(
     organization := "com.mfglabs"
   , version := "0.4.6-rc3"
   , isSnapshot := false
-  , crossScalaVersions := Seq("2.11.12", "2.12.6")
+  , crossScalaVersions := Seq("2.11.12", "2.12.8")
```
Comment (martin1keogh) [inline build.sbt]: FYI 2.12.9 came out a couple of days ago (cf https://www.scala-lang.org/news/2.12.9)

File: build.sbt [hunk 2]
```diff
@@ -153,12 +179,17 @@ lazy val influx =
...
+      libraryDependencies += "org.influxdb" % "influxdb-java" % "2.15"
+    )
+    .dependsOn(core)
+
+lazy val applicationinsights =
+  project.in(file("precepte-applicationinsights"))
+    .settings(commonSettings:_*)
+    .settings(
+      name := "precepte-applicationinsights",
+      libraryDependencies += "com.microsoft.azure" % "applicationinsights-core" % "2.4.0-BETA"
```
Comment (martin1keogh) [inline build.sbt]: 2.4.0 is out of beta btw (cf https://mvnrepository.com/artifact/com.microsoft.azure/applicationinsights-core/2.4.1)

File: precepte-core-cats/src/main/scala/HackSI2712.scala [hunk 3]
```diff
@@ -48,25 +48,33 @@ trait PrecepteHackSI2712[TCA, TC[_[_], _], M[_[_]], F[_], Ta, MS, UMS, A] {
 object PrecepteHackSI2712 {
```
Comment (martin1keogh) [inline precepte-core-cats/src/main/scala/HackSI2712.scala]: Is this still needed? Didn't the SI2712 fix get backported to scala 2.11.9?

File: precepte-core-cats/src/main/scala/package.scala [hunk 4]
```diff
@@ -12,62 +12,114 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+        with cats.Defer[Precepte[T, M, U, F, ?]] {
+
+      final type precepte[A] = Precepte[T, M, U, F, A]
+      final type instrumentStep = SubStepInstrumentation[T, M, U, F]
+      final type state = PState[T, M, U]
+
+      final val monad: Monad[Precepte[T, M, U, F, ?]] = this
+
+      @inline final def inspect[A](
+          f: PState[T, M, U] => A): Precepte[T, M, U, F, A] =
```
Comment (martin1keogh) [inline precepte-core-cats/src/main/scala/package.scala]:
```suggestion
          f: PState[T, M, U] => A): precepte[A] =
```
?

Shouldn't change anything, mostly for symmetry with the `def`s below.

File: precepte-core-cats/src/main/scala/package.scala [hunk 5]
```diff
@@ -12,62 +12,114 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+      @inline final def flatMap[A, B](fa: precepte[A])(
+          f: A => precepte[B]): precepte[B] =
+        fa.flatMap(f)
+      @inline final override def ap[A, B](ff: precepte[A => B])(
+          fa: precepte[A]): precepte[B] =
+        fa.ap(ff)
+
+      @inline final def tailRecM[A, B](a: A)(
+          f: A => Precepte[T, M, U, F, Either[A, B]])
+        : Precepte[T, M, U, F, B] = {
```
Comment (martin1keogh) [inline precepte-core-cats/src/main/scala/package.scala]:
```suggestion
        : precepte[B] = {
```

File: precepte-core-cats/src/main/scala/package.scala [hunk 6]
```diff
@@ -12,62 +12,114 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+
+      @inline final def flatMap[A, B](fa: precepte[A])(
+          f: A => precepte[B]): precepte[B] =
+        fa.flatMap(f)
+      @inline final override def ap[A, B](ff: precepte[A => B])(
+          fa: precepte[A]): precepte[B] =
+        fa.ap(ff)
+
+      @inline final def tailRecM[A, B](a: A)(
+          f: A => Precepte[T, M, U, F, Either[A, B]])
```
Comment (martin1keogh) [inline precepte-core-cats/src/main/scala/package.scala]:
```suggestion
          f: A => precepte[Either[A, B]])
```

File: precepte-core-cats/src/main/scala/package.scala [hunk 7]
```diff
@@ -77,26 +129,34 @@ package object corecats extends SubMeta {
...
-  def future[Ta, M, U, A](ta: Ta)(λ: Precepte[Ta, M, U, Future, A]#S => Future[A])(implicit func: Functor[Future], ec: scala.concurrent.ExecutionContext): Precepte[Ta, M, U, Future, Either[Throwable, A]] =
-    Precepte(ta){ pa =>
-      func.map(λ(pa))(Right(_))
-        .recover{ case e => Left(e) }
+  implicit def xorHasHoist[A]
+    : HasHoist.Aux[Either[A, ?], λ[(α[_], β) => EitherT[α, A, β]]] =
+    new XorHasHoist[A]
+
+  @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
+  def future[Ta, M, U, A](ta: Ta)(λ: PState[Ta, M, U] => Future[A])(
```
Comment (martin1keogh) [inline precepte-core-cats/src/main/scala/package.scala]: I know this wasn't modified by this PR but IMO either use `tag: Tag` or stick with `t: T` like all the others `Precepte[T, M, U, F, A]`

File: precepte-core-cats/src/test/scala/PrecepteSpec.scala [hunk 8]
```diff
@@ -532,39 +577,104 @@ class PrecepteSpec extends FlatSpec with ScalaFutures with Inside {
...
+    def g(i: Long): P.precepte[Int] =
+      if (i <= 0) P.pure(0)
+      else
+        for {
+          x <- f(i)
+          q <- g(i - 1)
+        } yield x + q
+
+    val n = 150L
+    val exceptions =
```
Comment (martin1keogh) [inline precepte-core-cats/src/test/scala/PrecepteSpec.scala]: Maybe add a comment saying "numbers divisible by either 3 or 5 will raise a DivisionByZero in f(), meaning they'll count as '1' (see the `recoverWith`). We expect `g()`'s return value to be the same as the number of thrown exceptions"

File: precepte-core/src/main/scala/adts.scala [hunk 9]
```diff
@@ -12,32 +12,80 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
-  def mapUnmanaged[UnmanagedState2](f: UnmanagedState => UnmanagedState2): PState[Ta, ManagedState, UnmanagedState2] = PState(managed, f(unmanaged))
+  def mapUnmanaged[UnmanagedState2](f: UnmanagedState => UnmanagedState2)
+    : PState[Ta, ManagedState, UnmanagedState2] = PState(managed, f(unmanaged))
 }
 
 /** A Typeclass representing an updatable Precepte state (can append a Tag+idx & update unmanaged part) */
-trait PStateUpdater[Ta, MS, FS] {
-  type S = PState[Ta, MS, FS]
-  def appendTags(s: S, t: Ta, idx: Int): S
+trait PStateUpdater[Ta, MS, FS] { self =>
```
Comment (martin1keogh) [inline precepte-core/src/main/scala/adts.scala]: Why `FS` here? (instead of the usual `U`)

File: build.sbt (hunk 1 above)
Comment (chrilves) [reply build.sbt]: Thanks. Unfortunately, the splain plugin is not yet available for 2.12.9.

File: build.sbt (hunk 2 above)
Comment (chrilves) [reply build.sbt]: :+1:

File: precepte-core-cats/src/main/scala/HackSI2712.scala (hunk 3 above)
Comment (chrilves) [reply precepte-core-cats/src/main/scala/HackSI2712.scala]: Probably, but i don't want to break code depending on it until i'm sure it is useless.

File: precepte-core/src/main/scala/adts.scala (hunk 9 above)
Comment (chrilves) [reply precepte-core/src/main/scala/adts.scala]: I don't know, was there before.

**Pull request merged**